daemon/pool.go

```diff
@@ -73,123 +100,445 @@ func NewDriverPool(factory FactoryFunction) *DriverPool {
+func (dp *DriverPool) spawnOne() {
+	dp.spawning.Add(1)
+	defer dp.spawning.Add(-1)
+
+	// TODO(dennwc): use exponential backoff instead?
+	ticker := time.NewTicker(time.Millisecond * 250)
+	defer ticker.Stop()
+
+	// keep trying in case of a failure
+	for {
+		d, err := dp.factory(chanContext(dp.stop))
+		if err == nil {
+			dp.drivers.Lock()
+			dp.drivers.all[d] = struct{}{}
+			dp.running.Add(1)
+			dp.drivers.Unlock()
+
+			err = dp.putDriver(d)
+			if err == nil {
+				return // done
+			}
+		}
+		dp.Logger.Errorf("failed to start a driver: %v", err)
+		select {
+		case <-dp.stop:
+			return // cancel
+		case dp.spawnErr <- err:
+		case <-ticker.C:
+		}
 	}
-	return dp.del(-dn)
 }
 
-func (dp *DriverPool) add(ctx context.Context, n int) error {
-	for i := 0; i < n; i++ {
-		d, err := dp.factory(ctx)
-		if err != nil {
-			return err
+// runSpawn is a goroutine responsible for spawning new instances in the background.
+func (dp *DriverPool) runSpawn() {
+	defer dp.wg.Done()
+	for {
+		select {
+		case <-dp.stop:
+			return
+		case <-dp.spawn:
+			dp.spawnOne()
 		}
+	}
+}
 
-		dp.drivers.Lock()
-		dp.drivers.byID[d.ID()] = d
-		dp.drivers.Unlock()
-		dp.queue.Put(d)
-		dp.stats.instances.Add(1)
+// peekIdle tries to get an idle driver from the pool. It won't wait for the driver to
+// become idle, instead it will return false if there are no idle drivers.
+func (dp *DriverPool) peekIdle() (Driver, bool) {
+	dp.drivers.RLock()
+	n := len(dp.drivers.idle)
+	dp.drivers.RUnlock()
+	if n == 0 {
+		return nil, false
 	}
+	dp.drivers.Lock()
+	defer dp.drivers.Unlock()
+	for d := range dp.drivers.idle {
+		delete(dp.drivers.idle, d)
+		return d, true
+	}
+	return nil, false
+}
 
-	return nil
+// setIdle returns the driver to an idle state.
+func (dp *DriverPool) setIdle(d Driver) {
+	dp.drivers.Lock()
+	defer dp.drivers.Unlock()
+	dp.drivers.idle[d] = struct{}{}
 }
 
-func (dp *DriverPool) del(n int) error {
-	for i := 0; i < n; i++ {
-		d, more := dp.queue.Get()
-		if !more {
-			return ErrPoolClosed.New()
+// killDriver stops are removes the driver from the queue.
+func (dp *DriverPool) killDriver(d Driver) {
+	dp.drivers.Lock()
+	delete(dp.drivers.all, d)
+	delete(dp.drivers.idle, d)
+	dp.running.Add(-1)
+	dp.exited.Add(1)
+	dp.drivers.Unlock()
+
+	if err := d.Stop(); err != nil {
+		dp.Logger.Errorf("error removing stopped driver: %s", err)
+	}
+}
+
+// scaleDiff returns current difference between the target number of instances and the
+// current number of running instances.
+func (dp *DriverPool) scaleDiff() int {
+	total := dp.running.Value()
+	return dp.target.Value() - total
+}
+
+// scale the driver pool to the current target number of instances.
+func (dp *DriverPool) scale() {
+	dn := dp.scaleDiff()
+	if dn == 0 {
+		return
+	}
+
+	// rescaleLater interrupts the scaling and serves the request first.
+	// It will make sure to continue scaling later.
+	rescaleLater := func(req driverRequest) {
```

Comment:
Can this be a method of `*DriverPool`? It doesn't seem to close over anything else in the environment.